### ISSUE TITLE
atomic_shared_ptr.h doesn't compile with gcc-9.3 / C++20

### ISSUE THREAD

**Issue by maxgolov** [open-telemetry/opentelemetry-cpp#105]
Implementation suggests that it's C++20 compatible here:
https://github.com/open-telemetry/opentelemetry-cpp/blob/9e5924c523c3cd36ecbc38be1c93b6bb162a84d3/sdk/include/opentelemetry/sdk/common/atomic_shared_ptr.h#L14

However, if the project is built with
```
set(CMAKE_CXX_STANDARD 20)
```

Then the build is broken here:
https://github.com/open-telemetry/opentelemetry-cpp/blob/9e5924c523c3cd36ecbc38be1c93b6bb162a84d3/sdk/include/opentelemetry/sdk/trace/tracer.h#L47

With the following error:
```
[ 50%] Building CXX object sdk/src/trace/CMakeFiles/opentelemetry_trace.dir/tracer_provider.cc.o
In file included from /mnt/c/work/opentelemetry-cpp/sdk/include/opentelemetry/sdk/common/atomic_shared_ptr.h:3,
                 from /mnt/c/work/opentelemetry-cpp/sdk/include/opentelemetry/sdk/trace/tracer.h:3,
                 from /mnt/c/work/opentelemetry-cpp/sdk/include/opentelemetry/sdk/trace/tracer_provider.h:9,
                 from /mnt/c/work/opentelemetry-cpp/sdk/src/trace/tracer_provider.cc:1:
/usr/include/c++/9/atomic: In instantiation of ‘struct std::atomic<std::shared_ptr<opentelemetry::v0::sdk::trace::SpanProcessor> >’:
/mnt/c/work/opentelemetry-cpp/sdk/include/opentelemetry/sdk/common/atomic_shared_ptr.h:33:35:   required from ‘class opentelemetry::v0::sdk::AtomicSharedPtr<opentelemetry::v0::sdk::trace::SpanProcessor>’
/mnt/c/work/opentelemetry-cpp/sdk/include/opentelemetry/sdk/trace/tracer.h:47:54:   required from here
/usr/include/c++/9/atomic:191:21: error: static assertion failed: std::atomic requires a trivially copyable type
  191 |       static_assert(__is_trivially_copyable(_Tp),
```

**Comment by maxgolov** [comment]
I was evaluating what happens when we build on latest Ubuntu release with its stock compiler in a project that is targeting C++20. The build is not gonna succeed.

**Comment by pyohannes** [comment]
I can look into this. The related template specialization is [supposed to be part of C++20](https://en.cppreference.com/w/cpp/memory/shared_ptr/atomic2).

**Comment by maxgolov** [comment]
Fixed by #114
